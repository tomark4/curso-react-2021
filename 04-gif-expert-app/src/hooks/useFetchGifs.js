import {useState, useEffect} from 'react'
import {getGifs} from '../helpers/getGifs'

export const useFetchGifs = (category) => {
  const [gifs, setGifs] = useState({
    data: [],
    loading: true
  });

  useEffect(() => {
    getGifs(category)
    .then( imgs => setGifs({loading: false, data: imgs}))
    .catch(err => console.log(err));
  }, [category]);

  return gifs;
}
