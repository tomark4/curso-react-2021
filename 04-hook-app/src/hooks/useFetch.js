import { useRef, useEffect, useState } from "react"

const useFetch = (url) => {
  const isMounted = useRef(true);

  const [state, setState] = useState({data: [], loading: true, error: null});

  useEffect(() => {
    return () => { isMounted.current = false; }
  },[]);

  useEffect(() => {
    setState({data:[],error: null, loading: true});

    fetch(url)
    .then( resp => resp.json() )
    .then( data => {
      if(isMounted.current){
        setState({
          data,
          loading: false,
          error: null,
        })
      }
    })
    .catch( err => {
      setState({
        data: [],
        loading: false,
        error: err
      })
    });
  },[url])

  return state;
}

export default useFetch
