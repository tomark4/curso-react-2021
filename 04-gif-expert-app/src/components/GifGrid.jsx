// import React, { useEffect, useState } from "react";
import React from "react";
import PropTypes from "prop-types";
import { useFetchGifs } from "../hooks/useFetchGifs";
import GifGridItem from "./GifGridItem";

const GifGrid = ({ name }) => {


  const {data: imgs, loading} = useFetchGifs(name);

  return (
    <>
      <h3>{name}</h3>
      <div>
        { loading && <p>'Cargando...'</p>}
        <ul>
            {imgs.map((img) => (
            <GifGridItem key={img.id} {...img} />
            ))}
        </ul>
        <div style={{ clear: 'both'}}></div>
      </div>
    </>
  );
};

GifGrid.propTypes = {
  name: PropTypes.string.isRequired,
};

export default GifGrid;
