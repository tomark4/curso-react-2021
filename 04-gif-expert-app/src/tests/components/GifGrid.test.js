import { shallow } from "enzyme";
import GifGrid from '../../components/GifGrid';
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en GifGrid component', () => {

  test('should be match with snapshot', () => {
    useFetchGifs.mockReturnValue({
      data: [] ,
      loading: true
    })
    const wrapper = shallow( <GifGrid name="goku"/>)
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de mostrar items cuando se cargan imagenes ', () => {
    const imgs = [
      {id: 'abc',url: 'http:/localhost/img.jpg',title: 'whatever'},
      {id: '123',url: 'http:/localhost/img.jpg',title: 'whatever'}
    ];

    useFetchGifs.mockReturnValue({
      data: imgs,
      loading: false
    });

    const wrapper = shallow( <GifGrid name="goku"/>)

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find('p').exists() ).toBe(false);
    expect( wrapper.find('GifGridItem').length ).toBe(imgs.length);
  });


});
