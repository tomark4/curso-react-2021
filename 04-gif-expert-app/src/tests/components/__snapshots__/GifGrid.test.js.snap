// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifGrid component debe de mostrar items cuando se cargan imagenes  1`] = `
<Fragment>
  <h3>
    goku
  </h3>
  <div>
    <ul>
      <GifGridItem
        id="abc"
        key="abc"
        title="whatever"
        url="http:/localhost/img.jpg"
      />
      <GifGridItem
        id="123"
        key="123"
        title="whatever"
        url="http:/localhost/img.jpg"
      />
    </ul>
    <div
      style={
        Object {
          "clear": "both",
        }
      }
    />
  </div>
</Fragment>
`;

exports[`Pruebas en GifGrid component should be match with snapshot 1`] = `
<Fragment>
  <h3>
    goku
  </h3>
  <div>
    <p>
      'Cargando...'
    </p>
    <ul />
    <div
      style={
        Object {
          "clear": "both",
        }
      }
    />
  </div>
</Fragment>
`;
